### PR TITLE
Implement layered OSL closures

### PR DESCRIPTION
**Pull request by est77**

### CODE REVIEW

File: src/appleseed/CMakeLists.txt [hunk 1]
```diff
@@ -1632,6 +1632,8 @@ set (renderer_modeling_bsdf_sources
     renderer/modeling/bsdf/glassbsdf.h
     renderer/modeling/bsdf/glossybrdf.cpp
     renderer/modeling/bsdf/glossybrdf.h
+    renderer/modeling/bsdf/glossylayerbsdf.cpp
```
Comment (dictoon) [inline src/appleseed/CMakeLists.txt:1635]: I assume that not exposing this BSDF in the API header files is intentional?

File: src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+                alpha_x,
+                alpha_y);
+
+            const FresnelDielectricFun f(
+                values->m_reflectance,
+                1.0f,
+                values->m_precomputed.m_outside_ior / values->m_ior,
+                1.0f);
+
+            MicrofacetBRDFHelper<GGXMDF, true>::sample(
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp:166]: Why is Flip being set to `true` here (and in the Glossy BRDF as well)?

File: src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+        {
+            return Model;
+        }
+
+        size_t compute_input_data_size() const override
+        {
+            return sizeof(InputValues);
+        }
+
+        void prepare_inputs(
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp:124]: The Glossy BRDF does this:

    values->m_roughness = std::max(values->m_roughness, shading_point.get_ray().m_min_roughness);

Isn't it necessary here too for roughness clamping to work?

File: src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+            const float Eo = get_dielectric_layer_directional_albedo(
+                eta,
+                values->m_roughness,
+                abs(cos_on));
+            const float Ei = get_dielectric_layer_directional_albedo(
+                eta,
+                values->m_roughness,
+                abs(cos_in));
+
+            value *= min(1.0f - Ei, 1.0f - Eo);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp:361]: `std::min()`

File: src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+            {
+                const float Lo = safe_rcp(abs(cos_on), 1e-6f);
+                const float Li = safe_rcp(abs(cos_in), 1e-6f);
+                float L = Li + Lo;
+
+                // For transmission, use the average length.
+                if (cos_in * cos_on <= 0.0f)
+                    L *= 0.5f;
+
+                value *= pow(values->m_transmittance, values->m_thickness * L);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp:373]: `std::pow()` (add `std::` prefix to all standard functions)

File: src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+
+            const Vector3f& normal = shading_basis.get_normal();
+            const float cos_on = dot(normal, outgoing);
+            const float cos_in = dot(normal, incoming);
+
+            const float eta = 1.0f / values->m_ior;
+            const float Eo = get_dielectric_layer_directional_albedo(
+                eta,
+                values->m_roughness,
+                abs(cos_on));
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp:355]: `std::abs()`

File: src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+            const float Ei = get_dielectric_layer_directional_albedo(
+                eta,
+                values->m_roughness,
+                abs(cos_in));
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp:359]: `std::abs()`

File: src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+#include <cstddef>
+#include <string>
+
+// Forward declarations.
+namespace foundation    { class IAbortSwitch; }
+namespace renderer      { class Assembly; }
+namespace renderer      { class Project; }
+
+using namespace foundation;
+using namespace std;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp:72]: We don't allow `using namespace std` anymore.

File: src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp [hunk 9]
```diff
@@ -0,0 +1,513 @@
...
+      public:
+        GlossyLayerAlbedoTable(const float* table, const float min_eta, const float max_eta)
+          : AlbedoTable3D(table, min_eta, max_eta)
+        {
+        }
+
+        GlossyLayerAlbedoTable(const float min_eta, const float max_eta)
+          : AlbedoTable3D(min_eta, max_eta)
+        {
+            for (size_t z = 0; z < TableSize; ++z)
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp:395]: Use `std::size_t` in new code (fix everywhere).

File: src/appleseed/renderer/modeling/bsdf/bsdf.h [hunk 10]
```diff
@@ -198,6 +198,21 @@ class APPLESEED_DLLSYMBOL BSDF
...
 
+    // For layered BSDFs, compure the attenuation of substrate layers.
+    virtual void attenuate_substrate(
+        const void*                 data,                       // input values
+        const foundation::Basis3f&  shading_basis,              // world space orthonormal basis around shading normal
+        const foundation::Vector3f& outgoing,                   // world space outgoing direction, unit-length
+        const foundation::Vector3f& incoming,                   // world space incoming direction, unit-length
+        Spectrum&                   value) const;               // attenuated value
+
+    virtual void attenuate_substrate(
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/bsdf.h:209]: Why do we need to expose two variants of this method?

File: src/appleseed/renderer/modeling/bsdf/bsdf.cpp [hunk 11]
```diff
@@ -97,25 +97,45 @@ void* BSDF::evaluate_inputs(
...
 }
 
+void BSDF::attenuate_substrate(
+    const void*                 data,
+    const Basis3f&              shading_basis,
+    const Vector3f&             outgoing,
+    const Vector3f&             incoming,
+    Spectrum&                   value) const
+{
+    assert(false);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/bsdf.cpp:128]: Maybe this should be a bit more descriptive. If I understand correctly, we should never reach this assertion unless a BRDF that wasn't designed for layering is used, right? If so, what about something along these lines:

    assert(!"This BSDF does not support layering.");

File: src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp [hunk 12]
```diff
@@ -0,0 +1,513 @@
...
+            Vector3f m = GGXMDF::sample(wo, s, alpha);
+
+            const float cos_oh = std::abs(dot(wo, m));
+            const float cos_on = std::abs(wo.y);
+
+            if (cos_on == 0.0f || cos_oh == 0.0f)
+                return 0.0f;
+
+            const Vector3f n(0.0f, 1.0f, 0.0f);
+            Vector3f wi = reflect(wo, m);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp:461]: Make const.

File: src/appleseed/renderer/kernel/shading/closures.h [hunk 13]
```diff
@@ -217,17 +226,6 @@ class APPLESEED_ALIGN(16) CompositeSurfaceClosure
         const size_t                num_closures,
         float                       pdfs[MaxClosureEntries]) const;
 
-    void add_ior(
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.h:220]: Do we still need the `CompositeSurfaceClosure` class?

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 14]
```diff
@@ -702,6 +701,7 @@ namespace
                     p->T,
                     arena);
 
+            const float fresnel_weight = saturate(p->fresnel_weight);
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp:695]: Why did you move this? It looked alright where it was.

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 15]
```diff
@@ -1744,6 +1744,123 @@ namespace
...
+                composite_closure.add_closure<GlossyLayerBSDFInputValues>(
+                    id(),
+                    shading_basis,
+                    weight,
+                    p->N,
+                    p->T,
+                    arena);
+
+            const float roughness = clamp(p->roughness, 0.01f, 1.0f);
+            const float ior = max(p->ior, 0.001f);
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp:1830]: `std::max()`

File: src/appleseed/renderer/kernel/shading/closures.h [hunk 16]
```diff
@@ -117,10 +122,54 @@ enum ClosureID
...
-      : foundation::Exception(what)
+    explicit ExceptionOSLRuntimeError(const char* what);
+};
+
+
+//
+// Small closure layer ID stack.
+//
+
+template <size_t N>
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.h:133]: `std::size_t` (everywhere)

File: src/appleseed/renderer/kernel/shading/closures.h [hunk 17]
```diff
@@ -117,10 +122,54 @@ enum ClosureID
...
+
+    std::int8_t operator[](const size_t i) const
+    {
+        assert(i < m_size);
+        return m_ids[i];
+    }
+
+    void push(const std::int8_t id)
+    {
+        assert(m_size < N - 1);
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.h:160]: Shouldn't it be `m_size < N`?

File: src/appleseed/renderer/kernel/shading/closures.h [hunk 18]
```diff
@@ -117,10 +122,54 @@ enum ClosureID
...
+    }
+
+    void pop()
+    {
+        assert(m_size > 0);
+        --m_size;
+    }
+
+  private:
+    std::int8_t   m_ids[N];
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.h:171]: Shouldn't we use `std::uint8_t` instead? Or do we need to allow negative IDs?

File: src/appleseed/renderer/modeling/bsdf/oslbsdf.cpp [hunk 19]
```diff
@@ -402,12 +418,38 @@ namespace
...
+            const CompositeClosure&     c,
+            const size_t                closure_index,
+            const Vector3f&             outgoing,
+            const Vector3f&             incoming,
+            DirectShadingComponents&    value) const
+        {
+            const std::int8_t* layers = c.get_closure_layers(closure_index);
+            for (size_t i = 0; i < CompositeClosure::MaxClosureLayers; ++i)
+            {
+                if (layers[i] < 0)
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/oslbsdf.cpp:438]: I think I have my answer...

File: src/appleseed/renderer/modeling/bssrdf/oslbssrdf.cpp [hunk 20]
```diff
@@ -240,11 +270,42 @@ namespace
...
+                    outgoing,
+                    incoming,
+                    value);
+            }
+        }
+
+        const BSDF* bsdf_from_closure_id(const ClosureID cid) const
+        {
+            assert(cid >= FirstLayeredClosure);
+            return m_glossy_layer_bsdf.get();
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bssrdf/oslbssrdf.cpp:306]: Out of consistency with `bssrdf_from_closure_id()`, return a const ref instead of a const pointer (or change `bssrdf_from_closure_id()` to return a const pointer).

File: src/appleseed/CMakeLists.txt (hunk 1 above)
Comment (est77) [reply src/appleseed/CMakeLists.txt:1635]: Yes. It only works when using OSL.

File: src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp (hunk 5 above)
Comment (est77) [reply src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp:373]: Transmitance is a Spectrum here.

File: src/appleseed/renderer/modeling/bsdf/bsdf.h (hunk 10 above)
Comment (est77) [reply src/appleseed/renderer/modeling/bsdf/bsdf.h:209]: One is for BSDFs, the other one is used for BSSRDFs and EDFs.

File: src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp (hunk 12 above)
Comment (est77) [reply src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp:461]: force_above_surface(), a bit later modifies wi.

File: src/appleseed/renderer/kernel/shading/closures.h (hunk 18 above)
Comment (est77) [reply src/appleseed/renderer/kernel/shading/closures.h:171]: negative numbers are used to signal that there are no layers.

File: src/appleseed/renderer/kernel/shading/closures.h (hunk 13 above)
Comment (est77) [reply src/appleseed/renderer/kernel/shading/closures.h:220]: Yes. It has the closure tree traversal for BSDFs.
